ext/standard/string.c

```diff
@@ -3863,6 +3878,55 @@ PHPAPI zend_string *php_addcslashes_str(const char *str, size_t len, const char
 }
 /* }}} */
 
+static const char *get_char_repr(char c) {
+    for (size_t i = 0; i < sizeof(char_reprs) / sizeof(char_reprs[0]); i++) {
+        if (char_reprs[i].c == c) {
+            return char_reprs[i].repr;
+        }
+    }
+    return NULL;
+}
+
+/* {{{ php_repr_str */
+PHPAPI zend_string *php_repr_str(const char *str, size_t len) {
+    size_t new_len = 0;
+    char *repr_str;
+    
+    for (size_t i = 0; i < len; i++) {
+        const char *repr = get_char_repr(str[i]);
+        if (repr) {
+            new_len += strlen(repr);
+        } else if (isprint((unsigned char)str[i])) {
+            new_len += 1;
+        } else {
+            new_len += 4;
+        }
+    }
+    
+    repr_str = emalloc(new_len + 1);
```

Comment:
Minor suggestion:
Now you do emalloc+zend_string_init. You can already allocate a zend_string here by using zend_string_alloc. This will give you a `zend_string *` and you can then use `ZSTR_VAL` to write the characters directly into here. It avoids copying the string.

Reply:
Cool! That's a great idea, I would like to allocate the maximum required size and then shrink it after the encoding process just like what `php_addcslashes` does.

Reply:
Done, thank you for your feedback! Just like this reply (https://github.com/php/php-src/pull/15730#discussion_r1742659604). I have done some performance benchmarking, it turns out that the new version outperforms the old version by ~32% in CPU time.

Reply:
https://github.com/php/php-src/pull/15730/files#diff-fcf8a2a38ee4a0e3e2cb7c47251c9920ba8c5886d85969f676f9ddbee7aba503R3869-R3897